src/SFA.DAS.Commitments.Application/Commands/RejectTransferRequest/RejectTransferRequestCommandHandler.cs

```diff
@@ -0,0 +1,129 @@
+        private readonly IApprenticeshipEventsPublisher _apprenticeshipEventsPublisher;
+        private readonly IMessagePublisher _messagePublisher;
+        private readonly CohortApprovalService _cohortApprovalService;
+        private readonly HistoryService _historyService;
+
+        public RejectTransferRequestCommandHandler(AbstractValidator<RejectTransferRequestCommand> validator,
+            ICommitmentRepository commitmentRepository, IApprenticeshipRepository apprenticeshipRepository,
+            IApprenticeshipOverlapRules overlapRules, ICurrentDateTime currentDateTime,
+            IApprenticeshipEventsList apprenticeshipEventsList,
+            IApprenticeshipEventsPublisher apprenticeshipEventsPublisher, IMediator mediator,
+            IMessagePublisher messagePublisher,
+            IHistoryRepository historyRepository)
+        {
+            _validator = validator;
+            _commitmentRepository = commitmentRepository;
+            _currentDateTime = currentDateTime;
+            _apprenticeshipEventsList = apprenticeshipEventsList;
+            _apprenticeshipEventsPublisher = apprenticeshipEventsPublisher;
+            _messagePublisher = messagePublisher;
+            _historyService = new HistoryService(historyRepository);
+
+            _cohortApprovalService = new CohortApprovalService(apprenticeshipRepository, overlapRules, currentDateTime,
+                commitmentRepository, apprenticeshipEventsList, apprenticeshipEventsPublisher, mediator);
+        }
+
+        protected override async Task HandleCore(RejectTransferRequestCommand command)
+        {
+            _validator.ValidateAndThrow(command);
+
+            var commitment = await _commitmentRepository.GetCommitmentById(command.CommitmentId);
+            if (commitment == null)
+            {
+                throw new ResourceNotFoundException();
+            }
+
+            CheckAuthorization(command, commitment);
+            CheckCommitmentStatus(commitment, command);
+
+            _historyService.TrackUpdate(commitment, CommitmentChangeType.TransferSenderRejection.ToString(), commitment.Id, null, CallerType.TransferSender, command.UserEmail, commitment.ProviderId, command.TransferSenderId, command.UserName);
+
+            await _commitmentRepository.SetTransferApproval(command.CommitmentId,
+                TransferApprovalStatus.TransferRejected, command.UserEmail, command.UserName);
+
+            await _commitmentRepository.ResetEditStatusToEmployer(command.CommitmentId);
+
+            await UpdateCommitmentObjectWithNewValues(commitment);
+
+            await Task.WhenAll(
+                 _cohortApprovalService.ResetApprenticeshipsAgreementStatuses(commitment),
+                 _historyService.Save(),
+                 PublishApprenticeshipUpdatedEvents(commitment),
+                 PublishRejectedMessage(command)
+            );
+        }
+
+        private async Task UpdateCommitmentObjectWithNewValues(Commitment commitment)
```

Comment:
This is method slightly different when approving or rejecting, but could be common as well